metronome/networking/src/io/iohk/metronome/networking/RemoteConnectionManager.scala

```diff
@@ -0,0 +1,339 @@
+      retryConfig: RetryConfig
+  ): F[Unit] = {
+
+    /** Observable is used here as streaming primitive as it has richer api than Iterant and have mapParallelUnorderedF
+      * combinator, which makes it possible to have multiple concurrent retry timers, which are cancelled when whole
+      * outer stream is cancelled
+      */
+    Observable
+      .repeatEvalF(connectionsToAcquire.poll)
+      .filterEvalF(request => connectionsHandler.isNewConnection(request.key))
+      .mapEvalF { connectionToAcquire =>
+        connectTo(encryptedConnectionProvider, connectionToAcquire)
+      }
+      .mapParallelUnorderedF(Integer.MAX_VALUE) {
+        case Left(failure) =>
+          //TODO add logging of failure
+          val failureToLog = failure.err
+          retryConnection(retryConfig, failure.connectionRequest).flatMap(
+            updatedRequest => connectionsToAcquire.offer(updatedRequest)
+          )
+        case Right(connection) =>
+          val newOutgoingConnections =
+            HandledConnection.outgoing(connection.encryptedConnection)
+          connectionsHandler.registerIfAbsent(newOutgoingConnections)
+
+      }
+      .completedF
+  }
+
+  /** Reads incoming connections in linear fashion and check if they are on cluster allowed list.
+    */
+  private def handleServerConnections[F[_]: Concurrent: TaskLift, K, M: Codec](
+      pg: EncryptedConnectionProvider[F, K, M],
+      connectionsHandler: ConnectionHandler[F, K, M],
+      clusterConfig: ClusterConfig[K]
+  ): F[Unit] = {
+    Iterant
+      .repeatEvalF(pg.incomingConnection)
+      .takeWhile(_.isDefined)
+      .map(_.get)
+      .collect { case Right(value) =>
+        value
+      }
+      .mapEval { encryptedConnection =>
+        clusterConfig.getIncomingConnectionServerInfo(
+          encryptedConnection.remotePeerInfo._1
+        ) match {
+          case Some(incomingConnectionServerAddress) =>
+            val handledConnection = HandledConnection.incoming(
+              incomingConnectionServerAddress,
+              encryptedConnection
+            )
+            connectionsHandler.registerIfAbsent(handledConnection)
+
+          case None =>
+            // unknown connection, just close it
+            encryptedConnection.close()
+        }
+      }
+      .completedL
+  }
+
+  def withCancelToken[F[_]: Concurrent, A](
+      token: Deferred[F, Unit],
+      ops: F[Option[A]]
+  ): F[Option[A]] =
+    Concurrent[F].race(token.get, ops).map {
+      case Left(()) => None
+      case Right(x) => x
+    }
+
+  class HandledConnectionFinisher[F[_]: Concurrent: Timer, K, M](
+      connectionsToAcquire: ConcurrentQueue[F, OutGoingConnectionRequest[K]],
+      retryConfig: RetryConfig
+  ) {
+    def finish(handledConnection: HandledConnection[F, K, M]): F[Unit] = {
+      retryConnection(
+        retryConfig,
+        OutGoingConnectionRequest.initial(
+          handledConnection.key,
+          handledConnection.serverAddress
```

Comment:
Cool, in theory this could be a place where we consult the Discovery module to ask if the known address of the key has perhaps changed.